### ISSUE TITLE
Use DNS for simple read operation to registry

### ISSUE THREAD

**Issue by jucrouzet** [npm/npm#9534]
Using DNS instead of getting the whole JSON of a package via HTTP/Couch for simple READ operations like version checking or getting tarball url would improve npm in many ways.

I've written the idea here : https://medium.com/@c2c/npm-should-use-dns-for-optimization-a3e60bc0350

Would it be useful to make a PR ?

**Comment by othiym23** [comment]
Thanks for taking the time to write your article, and for your enthusiasm for making npm better! Here are a few points in response:

1. Over the last month, our CDN has had about about a 97.1% average cache hit rate. This includes both package metadata and tarballs, and still involves the transfer of a substantial number of bytes between our registry servers and Fastly, but our servers are not sweating particularly hard.
2. When the npm CLI makes a request to Fastly and gets a 304 in response (which comprises the _vast_ majority of package metadata requests once you've installed a few packages on your local system), there's no further parsing of the response necessary. In fact, npm itself doesn't even see the response body -- the [caching registry client](https://github.com/npm/npm/blob/b064c40b16c2c2c318ada2f502069002ba074e2c/lib/cache/caching-client.js) will return the package metadata from the cache at that point.
3. DNS lookups are indeed small and aggressively cached. As a result, DNS traffic is also aggressively throttled and given a low quality of service priority. The number of queries that npm fires off for a typical install is large, and it would only take a small percentage of them getting dropped or queued to cause a significant degradation of user experience.
4. This all presumes that all of the information used by npm is all public, which isn't the case in the case of private registries, caching registry proxies, and private packages. It also presumes that the registry is immutable, by which I mean this is something that would cause a lot of problems for unpublishes and deprecations. These aren't things that can't be worked around, but doing so would add complexity and also increased network traffic in certain cases, which sort of defeats the purpose.

So, in short, we have an architecture that's showing some growing pains, but overall is working pretty well, is tractable for our ops team, and is pretty easy to understand and scale. Thanks again for the offer, but this is a piece of npm we're unlikely to change in the near future!
